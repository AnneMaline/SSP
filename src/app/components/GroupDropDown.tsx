"use client";
import React, { useState } from "react";

type Member = {
  id: string;
  role: string;
};

type GroupDropDownProps = {
  name: string;
  description: string;
  members: Member[];
  onDelete: () => void;
};

const GroupDropDown: React.FC<GroupDropDownProps> = ({
  name,
  description,
  members: initialMembers,
  onDelete,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const [members, setMembers] = useState<Member[]>(initialMembers);

  const toggleDropdown = () => {
    setIsOpen((prevState) => !prevState);
  };

  // Add member. Template - Change!!!!
  const handleAddMember = () => {
    const newId = (members.length + 1).toString();
    const newRole = `New Role ${newId}`;
    setMembers((prev) => [...prev, { id: newId, role: newRole }]);
  };

  return (
    <div className="border border-gray-300 mr-2 ml-2">
      {/* Dropdown Header */}
      <div
        className="flex justify-between items-center px-4 py-2 cursor-pointer"
        onClick={toggleDropdown}
      >
        <span>{name}</span>
        <span>{isOpen ? "↑" : "↓"}</span>
      </div>

      {/* Dropdown Body (visible when isOpen is true) */}
      {isOpen && (
        <div className="p-4">
          {/* Description */}
          <div className="mb-4 text-gray-600">
            <p>{description}</p>
          </div>

          {/* Table of Members */}
          <div className="mb-4">
            <p className="font-semibold text-lg mb-2">
              Number of members: {members.length}
            </p>
            <table className="w-full table-auto border-collapse border border-gray-300">
              <thead>
                <tr>
                  <th className="border border-gray-300 px-4 py-2">ID</th>
                  <th className="border border-gray-300 px-4 py-2">Role</th>
                </tr>
              </thead>
              <tbody>
                {members.map((member) => (
                  <tr key={member.id}>
                    <td className="border border-gray-300 px-4 py-2">
                      {member.id}
                    </td>
                    <td className="border border-gray-300 px-4 py-2">
                      {member.role}
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          </div>
          {/* Buttons */}
          <div className="flex justify-between mt-4">
            <button
              onClick={handleAddMember}
              className="px-4 py-2 bg-blue-500 text-white rounded hover:bg-blue-600"
            >
              Add Member
            </button>
            <button
              onClick={onDelete}
              className="px-4 py-2 bg-red-500 text-white rounded hover:bg-red-600"
            >
              Delete
            </button>
          </div>
        </div>
      )}
    </div>
  );
};

export default GroupDropDown;
