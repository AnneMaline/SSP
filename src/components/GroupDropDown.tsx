"use client";
import React, { useState } from "react";
import AddIDtoGroupForm from "./AddIDtoGroupForm";
import { validateAuth } from "@/utils/validateAuth";

type Member = {
  email: string;
  role: string;
};

type GroupDropDownProps = {
  name: string;
  group_email: string;
  description: string;
  data_partition_id: string;
};

const GroupDropDown = ({
  name,
  group_email,
  description,
  data_partition_id,
}: GroupDropDownProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const [members, setMembers] = useState<Member[]>([]);
  const [membersCount, setMembersCount] = useState<number>(0);
  const [showMembers, setShowMembers] = useState(false);
  const [showAddIDForm, setShowAddIDForm] = useState(false);

  const toggleDropdown = async () => {
    setIsOpen((prevState) => !prevState);
    if (!isOpen && membersCount === 0) {
      const authToken = await validateAuth();

      try {
        const response = await fetch(
          `/api/entitlements/v2/groups/${group_email}/membersCount`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
              "data-partition-id": data_partition_id,
              group_email: group_email,
              Authorization: `Bearer ${authToken}`,
            },
          }
        );
        if (!response.ok) {
          throw new Error(`Error: ${response.status} ${response.statusText}`);
        }
        const data = await response.json();
        setMembersCount(data.membersCount);
      } catch (error) {
        console.error("Error fetching members count:", error);
      }
    }
  };

  const handleSeeMembers = async () => {
    if (showMembers) {
      setShowMembers(false);
      return;
    } else {
      setShowMembers(true);
    }

    if (members.length == 0) {
      const authToken = await validateAuth();

      try {
        const response = await fetch(
          `/api/entitlements/v2/groups/${group_email}/members/getMembers`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
              "data-partition-id": data_partition_id,
              group_email: group_email,
              // role,
              // includeType,
              Authorization: `Bearer ${authToken}`,
            },
          }
        );
        if (!response.ok) {
          throw new Error(`Error: ${response.status} ${response.statusText}`);
        }
        const data = await response.json();
        setMembers(data.members);
      } catch (error) {
        console.error("Error fetching members:", error);
      }
    }
  };

  const handleAddMember = () => {
    setShowAddIDForm(true); // Show the AddIDtoGroupForm pop-up
  };

  return (
    <div className="border border-gray-300 mr-2 ml-2">
      {/* Dropdown Header */}
      <div
        className="flex justify-between items-center px-4 py-2 cursor-pointer"
        onClick={toggleDropdown}
      >
        <span>{name}</span>
        <span>{isOpen ? "↑" : "↓"}</span>
      </div>

      {/* Dropdown Body (visible when isOpen is true) */}
      {isOpen && (
        <div className="p-4">
          {/* Description */}
          <div className="mb-4 text-gray-600">
            <p>{description}</p>
          </div>

          {/* Number of Members */}
          <div className="mb-4">
            <p className="font-semibold text-lg mb-2">
              Number of members: {membersCount}
            </p>
          </div>

          {/* Buttons */}
          <div className="mb-4">
            <button
              className="bg-blue-500 text-white px-4 py-2 mr-2"
              onClick={handleSeeMembers}
            >
              {!showMembers ? "See Members" : "Hide Members"}
            </button>
            <button
              className="bg-blue-500 text-white px-4 py-2 mr-2"
              onClick={handleAddMember}
            >
              Add Member
            </button>
          </div>

          {/* Table of Members (visible when showMembers is true) */}
          {showMembers && (
            <div className="mb-4">
              <table className="w-full table-auto border-collapse border border-gray-300">
                <thead>
                  <tr>
                    <th className="border border-gray-300 px-4 py-2">Email</th>
                    <th className="border border-gray-300 px-4 py-2">Role</th>
                    <th className="border border-gray-300 px-4 py-2">
                      Actions
                    </th>
                  </tr>
                </thead>
                <tbody>
                  {members.map((member) => (
                    <tr key={member.email}>
                      <td className="border border-gray-300 px-4 py-2">
                        {member.email}
                      </td>
                      <td className="border border-gray-300 px-4 py-2">
                        {member.role}
                      </td>
                    </tr>
                  ))}
                </tbody>
              </table>
            </div>
          )}

          {/* AddIDtoGroupForm Pop-up */}
          {showAddIDForm && (
            <div className="fixed inset-0 flex items-center justify-center bg-black bg-opacity-50">
              <div className="bg-white p-4 rounded">
                <AddIDtoGroupForm />
                <button
                  className="mt-4 bg-gray-500 text-white px-4 py-2"
                  onClick={() => setShowAddIDForm(false)}
                >
                  Close
                </button>
              </div>
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default GroupDropDown;
