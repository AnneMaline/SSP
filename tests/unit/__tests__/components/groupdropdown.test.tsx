import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import GroupDropDown from "../../../../src/app/components/GroupDropDown";

describe("GroupDropDown Component", () => {
  const mockOnDelete = jest.fn();

  const mockProps = {
    name: "Test Group",
    description: "This is a test group.",
    members: [
      { id: "1", role: "Admin" },
      { id: "2", role: "Member" },
    ],
    onDelete: mockOnDelete,
  };

  it("renders correctly", () => {
    const { container } = render(<GroupDropDown {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders the component with correct initial state", () => {
    render(<GroupDropDown {...mockProps} />);

    // Check if the name is displayed in the header
    expect(screen.getByText("Test Group")).toBeInTheDocument();

    // Check if the dropdown is initially closed
    expect(screen.queryByText("Number of members: 2")).not.toBeInTheDocument();
    expect(screen.getByText("↓")).toBeInTheDocument();
  });

  it("toggles the dropdown on click", () => {
    render(<GroupDropDown {...mockProps} />);

    const header = screen.getByText("Test Group");

    // Open the dropdown
    fireEvent.click(header);
    expect(screen.getByText("Number of members: 2")).toBeInTheDocument();
    expect(screen.getByText("↑")).toBeInTheDocument();

    // Close the dropdown
    fireEvent.click(header);
    expect(screen.queryByText("Number of members: 2")).not.toBeInTheDocument();
    expect(screen.getByText("↓")).toBeInTheDocument();
  });

  it("displays the correct number of members", () => {
    render(<GroupDropDown {...mockProps} />);

    // Open the dropdown
    fireEvent.click(screen.getByText("Test Group"));

    // Check the number of members
    expect(screen.getByText("Number of members: 2")).toBeInTheDocument();

    // Check member table rows
    const rows = screen.getAllByRole("row");
    expect(rows).toHaveLength(3); // 1 header row + 2 member rows
  });

  it("adds a member when the 'Add Member' button is clicked", () => {
    render(<GroupDropDown {...mockProps} />);

    // Open the dropdown
    fireEvent.click(screen.getByText("Test Group"));

    // Click the Add Member button
    const addButton = screen.getByText("Add Member");
    fireEvent.click(addButton);

    // Verify the new member count
    expect(screen.getByText("Number of members: 3")).toBeInTheDocument();

    // Verify the new member row
    expect(screen.getByText("New Role 3")).toBeInTheDocument();
  });

  it("calls the onDelete function when the 'Delete' button is clicked", () => {
    render(<GroupDropDown {...mockProps} />);

    // Open the dropdown
    fireEvent.click(screen.getByText("Test Group"));

    // Click the Delete button
    const deleteButton = screen.getByText("Delete");
    fireEvent.click(deleteButton);

    // Verify the onDelete function is called
    expect(mockOnDelete).toHaveBeenCalledTimes(1);
  });
});
