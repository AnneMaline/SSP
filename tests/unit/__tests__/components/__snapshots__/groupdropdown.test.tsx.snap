// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupDropDown Component renders correctly 1`] = `
<div>
  <div
    class="border border-gray-300 mr-2 ml-2"
  >
    <div
      class="flex justify-between items-center px-4 py-2 cursor-pointer"
    >
      <span>
        Test Group
      </span>
      <span>
        â†“
      </span>
    </div>
  </div>
</div>
`;
